skglm/penalties/block_separable.py

```diff
@@ -335,19 +335,32 @@ def subdiff_distance(self, w, grad_ws, ws):
             w_g = w[grp_g_indices]
             norm_w_g = norm(w_g)
 
-            if self.positive and np.any(w_g < 0):
-                scores[idx] = np.inf
-            elif self.positive and norm_w_g == 0:
-                # distance of -norm(neg_grad_g) to weights[g] * [-alpha, alpha]
-                neg_grad_g = grad_g[grad_g < 0.]
-                scores[idx] = max(0, norm(neg_grad_g) - self.alpha * weights[g])
-            elif (not self.positive) and norm_w_g == 0:
-                # distance of -norm(grad_g) to weights[g] * [-alpha, alpha]
-                scores[idx] = max(0, norm(grad_g) - alpha * weights[g])
+            if norm_w_g == 0:
+                if self.positive:
+                    # distance of -neg_grad_g to weights[g] * [-alpha, alpha]
+                    neg_grad_g = grad_g[grad_g < 0.]
+                    scores[idx] = max(0, norm(neg_grad_g) -
+                                      self.alpha * weights[g])
+                else:
+                    # distance of -norm(grad_g) to weights[g] * [-alpha, alpha]
+                    scores[idx] = max(0, norm(grad_g) - alpha * weights[g])
             else:
-                # distance of -grad_g to the subdiff (here a singleton)
-                subdiff = alpha * weights[g] * w_g / norm_w_g
-                scores[idx] = norm(grad_g + subdiff)
+                if self.positive:
+                    if np.any(w_g < 0):
+                        scores[idx] = np.inf
+                    else:
+                        res = np.zeros_like(grad_g)
+                        for j in range(len(w_g)):
+                            thresh = alpha * weights[g] * w_g[j] / norm_w_g
+                            if w_g[j] > 0:
+                                res[j] = -grad_g[j] - thresh
+                            else:
+                                res[j] = max(-grad_g[j] - thresh, 0)
```

Comment:
In addition, I think I disagree with the formula, and the `max` might need to be a `min`

Reply:
Nice, how did you see it? It's weird that it's not tested, because if you inspect the solution of `test_GroupLasso_estimator`, it has some non zero groups, with both vanishing and non vanishing entries inside

Reply:
I have put a print and it gets tested.